include/seqlock.h

```diff
@@ -0,0 +1,348 @@
+ *
+ * In contrast to ACCESS_ONCE these two macros will also work on aggregate
+ * data types like structs or unions. If the size of the accessed data
+ * type exceeds the word size of the machine (e.g., 32 bits or 64 bits)
+ * READ_ONCE() and WRITE_ONCE()  will fall back to memcpy and print a
+ * compile-time warning.
+ *
+ * Their two major use cases are: (1) Mediating communication between
+ * process-level code and irq/NMI handlers, all running on the same CPU,
+ * and (2) Ensuring that the compiler does not  fold, spindle, or otherwise
+ * mutilate accesses that either do not require ordering or that interact
+ * with an explicit memory barrier or atomic instruction that provides the
+ * required ordering.
+ */
+#define READ_ONCE(x) \
+	({ union { typeof(x) __val; char __c[1]; } __u; \
+	__read_once_size(&(x), __u.__c, sizeof(x)); __u.__val; })
+
+#define cpu_relax() asm ("rep; nop" ::: "memory")
+
+/*
+ * Version using sequence counter only.
+ * This can be used when code has its own mutex protecting the
+ * updating starting before the write_seqcountbeqin() and ending
+ * after the write_seqcount_end().
+ */
+typedef struct seqcount {
+	unsigned sequence;
+} seqcount_t;
+
+static inline void
+seqcount_init(seqcount_t *s)
+{
+	s->sequence = 0;
+}
+
+/*
+ * __read_seqcount_begin - begin a seq-read critical section (without barrier).
+ * @s: pointer to seqcount_t
+ * Returns: count to be passed to read_seqcount_retry.
+ *
+ * __read_seqcount_begin is like read_seqcount_begin, but has no smp_rmb()
+ * barrier. Callers should ensure that smp_rmb() or equivalent ordering is
+ * provided before actually loading any of the variables that are to be
+ * protected in this critical section.
+ *
+ * Use carefully, only in critical code, and comment how the barrier is
+ * provided.
+ */
+static inline unsigned
+__read_seqcount_begin(const seqcount_t *s)
+{
+	unsigned ret;
+
+repeat:
+	ret = READ_ONCE(s->sequence);
+	if (unlikely(ret & 1)) {
+		cpu_relax();
+		goto repeat;
+	}
+	return ret;
+}
+
+/*
+ * raw_read_seqcount - Read the raw seqcount.
+ * @s: pointer to seqcount_t
+ * Returns: count to be passed to read_seqcount_retry.
+ *
+ * raw_read_seqcount opens a read critical section of the given
+ * seqcount without any lockdep checking and without checking or
+ * masking the LSB. Calling code is responsible for handling that.
+ */
+static inline unsigned
+raw_read_seqcount(const seqcount_t *s)
+{
+	unsigned ret = READ_ONCE(s->sequence);
+	rte_smp_rmb();
+	return ret;
+}
+
+/*
+ * raw_read_seqcount_begin - start seq-read critical section w/o lockdep.
+ * @s: pointer to seqcount_t
+ * Returns: count to be passed to read_seqcount_retry.
+ *
+ * raw_read_seqcount_begin opens a read critical section of the given
+ * seqcount, but without any lockdep checking. Validity of the critical
+ * section is tested by checking read_seqcount_retry function.
+ */
+static inline unsigned
+raw_read_seqcount_begin(const seqcount_t *s)
```

Comment:
The kernel has the `raw_*()` functions in this file because they have a debug mechanism for locks called lockdep. Since we don't have this mechanism in DPDK, we can merge the `raw_*()` functions into their single callers.